### PR TITLE
ModTool: Add features and update functionalities

### PR DESCRIPTION
**Pull request by swap-nil7**
This pull request contains the following features:-
1. A shift of CLI parsing from `Argparse` to `Click`. (additional dependency: `click`)
2. Plug-in Architecture for ModTool.
3. Support for external plug-ins. (Additional dependency: `click-plugins`)
4. Split of modules into `core`, `cli`, `templates`, `tests`, and `tools`.
5. Python API for ModTool
6. Tests for ModTool API
7. Logger for ModTool
8. ANSI colors to the logger stderr, ModTool Exception for CLI, input statements, ask_yes_or_no, etc. 
    (Optional Dependency: `colorama`)
9. Replacement of XML Generator by YAML Generator
10. Update of Templates (mainly addition of `yaml` default `grc` template)
11. Addition of XML to YAML converter as a command line utility as well as an API class.

Other features like `SequenceConpleter`, replacement of `raw open` with `with open(file) as f:` have been also added.

### CODE REVIEW

File: gr-utils/python/modtool/CMakeLists.txt [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Copyright 2011 Free Software Foundation, Inc.
+# Copyright 2018 Free Software Foundation, Inc.
```
Comment (primercuervo) [inline gr-utils/python/modtool/CMakeLists.txt]: When modifying a file, don't replace the year but add it to keep track of its changes

`# Copyright 2011, 2018 Free Software Foundation, Inc.`

This goes for all mods on the Copyright year that you made

File: gr-utils/python/modtool/cli/base.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+from pkg_resources import iter_entry_points
+from logging import Formatter, StreamHandler
+
+import click
+from click import ClickException
+from click_plugins import with_plugins
+
+from gnuradio import gr
+
+sys.tracebacklimit = 0
```
Comment (primercuervo) [inline gr-utils/python/modtool/cli/base.py]: I thought we agreed on removing this limit and debugging the actual backtrace that was coming up

File: gr-utils/python/modtool/cli/newmod.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+import re
+import os
+
+import click
+
+from gnuradio import gr
+from ..core import ModToolNewModule
+from .base import ModToolException
+from .base import common_params, run, cli_input
```
Comment (primercuervo) [inline gr-utils/python/modtool/cli/newmod.py]: Why two different lines to import from .base?

File: gr-utils/python/modtool/cli/rename.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+# the Free Software Foundation, Inc., 51 Franklin Street,
+# Boston, MA 02110-1301, USA.
+#
+""" Module to rename blocks """
+
+import re
+
+import click
+
+from ..core import ModToolRename
```
Comment (primercuervo) [inline gr-utils/python/modtool/cli/rename.py]: Same as above. Unless having them in a single line is longer than 80 chars, having them in a single line helps readability

File: gr-utils/python/modtool/core/add.py [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+            self.scm.mark_files_updated((self._file['cminclude'], self._file['cmlib']))
+
+    def _run_swig(self):
+        """ Do everything that needs doing in the subdir 'swig'.
+        - Edit main *.i file
+        """
+        if self._get_mainswigfile() is None:
+            logger.warning('Warning: No main swig file found.')
+            return
+        logger.info("Editing %s..." % self._file['swig'])
```
Comment (primercuervo) [inline gr-utils/python/modtool/core/add.py]: Using `'{}'.format(foo)` is preferred for string formatting

File: gr-utils/python/modtool/templates/gr-newmod/python/build_utils.py [hunk 6]
```diff
@@ -24,7 +24,9 @@
 from __future__ import absolute_import
 from __future__ import unicode_literals
 
-import re, os, os.path
+import re
+import os
+import os.path
```
Comment (primercuervo) [inline gr-utils/python/modtool/templates/gr-newmod/python/build_utils.py]: Is this necessary? seems to me that `import os` would be enough

File: gr-utils/python/modtool/templates/gr-newmod/python/build_utils_codes.py [hunk 7]
```diff
@@ -19,30 +18,29 @@
 # the Free Software Foundation, Inc., 51 Franklin Street,
 # Boston, MA 02110-1301, USA.
 #
+from __future__ import unicode_literals
```
Comment (primercuervo) [inline gr-utils/python/modtool/templates/gr-newmod/python/build_utils_codes.py]: is this necessary or are we still trying to get rid of all the future imports?

File: gr-utils/python/modtool/tests/test_modtool.py [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+        self.assertTrue(path.isdir(module_dir))
+        self.assertTrue(path.isdir(path.join(module_dir, 'lib')))
+        self.assertTrue(path.isdir(path.join(module_dir, 'python')))
+        self.assertTrue(path.isdir(path.join(module_dir, 'include')))
+        self.assertTrue(path.isdir(path.join(module_dir, 'docs')))
+        self.assertTrue(path.isdir(path.join(module_dir, 'cmake')))
+        self.assertTrue(path.isdir(path.join(module_dir, 'swig')))
+        self.assertTrue(path.exists(path.join(module_dir, 'CMakeLists.txt')))
+
+        ## pylint tests ##
```
Comment (primercuervo) [inline gr-utils/python/modtool/tests/test_modtool.py]: What is the purpose of this commented section? Is it usable if uncommented?

File: gr-utils/python/modtool/CMakeLists.txt (hunk 1 above)
Comment (swap-nil7) [reply gr-utils/python/modtool/CMakeLists.txt]: The files that are newly created will have only 2018. All the `core` scripts and the `tools` scripts have a previous copyright intact.

File: gr-utils/python/modtool/core/add.py (hunk 5 above)
Comment (swap-nil7) [reply gr-utils/python/modtool/core/add.py]: Oh! Yes, I just changed from `print` to `logging.logger`. I will amend it though.

File: gr-utils/python/modtool/tests/test_modtool.py (hunk 8 above)
Comment (swap-nil7) [reply gr-utils/python/modtool/tests/test_modtool.py]: Yes, it is usable but for me, it is highly time-consuming.

File: gr-utils/python/modtool/templates/gr-newmod/python/build_utils_codes.py (hunk 7 above)
Comment (swap-nil7) [reply gr-utils/python/modtool/templates/gr-newmod/python/build_utils_codes.py]: I did not remove future imports from the newmoddir. Shall I do it?

File: gr-utils/python/modtool/tests/test_modtool.py (hunk 8 above)
Comment (swap-nil7) [reply gr-utils/python/modtool/tests/test_modtool.py]: @primercuervo: Uncommented but now the tests may take as long as `12-13 secs`.

**Comment by swap-nil7** [comment]
@primercuervo: I have replaced `%s` or `%d` with `str.format()` everywhere except for some places where it was raising `KeyError`. I am done with the mentioned changes.

**Comment by marcusmueller** [comment]
sadly: 
<pre>
Traceback (most recent call last):
  File "/cache/workdir/Fedora_26_0/build_Fedora_26/build/python_compile_helper.py", line 6, in <module>
    py_compile.compile(file=src, cfile=gen, doraise=True)
  File "/usr/lib64/python2.7/py_compile.py", line 117, in compile
    raise py_exc
py_compile.PyCompileError:   File "/data/src/pull/1873/next/gr-utils/python/modtool/tests/test_modtool.py", line 106
    print(pylint_stdout.getvalue(), end='')
                                       ^
SyntaxError: invalid syntax
</pre>

**Comment by swap-nil7** [comment]
@marcusmueller: I thought the `next` was `Py3k-only`. Shall I import `print` and `unicode_literals` from `__future__`?

**Comment by noc0lour** [comment]
@swap-nil7 former `next` and soon-to-be `3.8` release is both Py2 and Py3k. Thus code should be compatible with both.

**Comment by primercuervo** [comment]
@marcusmueller @noc0lour   Well.. to be fair, that previous statement constitutes some big-deal miscommunication, as at the beginning of GSoC 2018 the work was agreed upon the `next` branch being Py3k only, and so the work was based on that agreement as well. I ignore how many changes are going to be needed for this PR to be compatible (it might as well just be one line), but I just wanted the record to state that this certainly modifies the (expectingly finished) roadmap for this work.

@swap-nil7 are you up to add this Py2k compatibility?

**Comment by swap-nil7** [comment]
I will try to make it version independent compatible but I don't really see the reason why Py2 compatibility is even needed.

**Comment by primercuervo** [comment]
@swap-nil7 thanks for taking care of this. Please do let us know if you need support in any way. 

The Py2k compatibility is for GNU Radio in general. If there is a component that requires still Py2k, then we can't just have other components (needless to say one so widely used as gr_modtool) that is Py3k only. We are trying to move all to Py3k, and we will at some point, but as of now it is just a decision to keep compatibility.

**Comment by primercuervo** [comment]
@swap-nil7 any update on this?

**Comment by swap-nil7** [comment]
@primercuervo: I apologize for the delay but I am not very sure about all the changes that should be incorporated for Py2k compatibility. As of now, I have just imported modules from `__future__`.

**Comment by noc0lour** [comment]
@swap-nil7 No problem, I'll take a look at what big chunks are missing :+1:

**Comment by noc0lour** [comment]
I have your latest branch rebased on master and will look for py2k compat

**Comment by noc0lour** [comment]
Two things we also need (aside from click) to get it running:
 - click-plugins
 - pyyaml with libyaml

The second one is quite hidden in CLoad/CDump calls. Maybe we can make CLoad/CDump optional if libyaml is available, that's how this is handled in GRC as well.

**Comment by noc0lour** [comment]
This should be all for Py2k compat: Basically only the last commit is needed for the compat. The other commits are fixing other issues on the way (which prohibited me from compiling/running gr_modtool): https://github.com/noc0lour/gnuradio/tree/modtool_rebase the branch is also already rebased on `master`.

**Comment by primercuervo** [comment]
@noc0lour thanks for taking care of this. In GRC what does it fall back to when libyaml isn't available? 

Also, in the commit including all the __future__ imports (i.e. https://github.com/gnuradio/gnuradio/pull/1873/commits/a9b88a46a01da0b7656c291c10ebf7be215f28e7 , https://github.com/noc0lour/gnuradio/commit/147f9f9462678099199fcb12cd74a914a9381cbc) are they _all_ needed in _all_ those files? I assume they are, but just wanted to ask out of curiosity.

Additionally, [PEP-8 isn't all that clear w.r.t multiple imports of the same module](https://www.python.org/dev/peps/pep-0008/#imports), but I believe these all future imports fit in a <80 chars one-liner.

**Comment by swap-nil7** [comment]
@noc0lour: Thanks a lot for your help. Please let me know if I can be of any help. :)

**Comment by noc0lour** [comment]
Sure thing. I personally really want to have new things and features (and big reworks in the source tree rather sooner than later :) ).
I actually don't know what would be best practice to move forward.
Maybe @marcusmueller could help with his opinion.
In practice you could either take my rebased branch and force-push it onto your branch or I can force-push it onto your branch. 
I'd rather not do that (if I accidently delete your history that would be a shame)'

**Comment by swap-nil7** [comment]
@marcusmueller: Why is the PR closed? :disappointed:

**Comment by marcusmueller** [comment]
@swap-nil7 that was completely unintentional. Github just did it because we've removed `next`, and I didn't re-target this PR to `master`. Sorry!

**Comment by hcab14** [comment]
I had a look at https://github.com/noc0lour/gnuradio/tree/modtool_rebase and indeed I was able to create an OOT with YAML definitions. There are two minor issues
- in CMakeLists.txt the gnuradio version has to be manually changed to 3.8
- the line `category: [tt]` should be `category: '[tt]'`, i.e., quotation marks have to be added to the template. Otherwise the category is interpreted as a python list.

Unfortunately I was not able to rebase https://github.com/noc0lour/gnuradio/tree/modtool_rebase to the current master.

**Comment by noc0lour** [comment]
Please continue discussion in #2284

**Pull request closed without merging**